### PR TITLE
fix: visit procedure declaration after visiting func/subrout in `ast_symboltable_visitor`

### LINKED ISSUE #5757: bug: defining procedure pointer with `subroutine` not allowed
### Minimal Reproducible Example

```fortran
program main

   procedure(say_something), pointer :: proc
   proc => say_something

   call proc("Hello from proc!")

contains

   subroutine say_something(message)
      character(len=*),intent(in) :: message
      print *, message
   end subroutine

end program main
```

### Output

```console
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran$ gfortran test.f90 && ./a.out
/usr/bin/ld: warning: /tmp/ccltcx46.o: requires executable stack (because the .note.GNU-stack section is executable)
 Hello from proc!
(base) saurabh-kumar@Awadh:~/Projects/System/lfortran$ lfortran test.f90
semantic error: Procedure type 'say_something' not declared
 --> test.f90:3:4
  |
3 |    procedure(say_something), pointer :: proc
  |    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ 


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

### PR DESCRIPTION
**Pull request by jinangshah21**
Fixes #5757

### CODE REVIEW

File: src/lfortran/semantics/ast_symboltable_visitor.cpp [hunk 1]
```diff
@@ -586,19 +586,43 @@ class SymbolTableVisitor : public CommonVisitor<SymbolTableVisitor> {
...
+                    if(type && type->m_type == AST::decl_typeType::TypeProcedure) {
+                        try {
+                            visit_unit_decl2(*x.m_decl[i]);
+                        } catch (SemanticAbort &e) {
+                            if ( !compiler_options.continue_compilation ) throw e;
+                        }
+                    }
+                }
+            }
+        }
```
Comment (kmr-srbh) [inline src/lfortran/semantics/ast_symboltable_visitor.cpp:624]: We have some redundancy here, also let's not loop over all declarations again. Can we add declarations with `AST::decl_typeType::TypeProcedure` to a vector `Vec<AST::unit_decl2_t>`, loop over them and visit?

Comment (jinangshah21) [reply src/lfortran/semantics/ast_symboltable_visitor.cpp:624]: Thanks ! I will do it.

File: src/lfortran/semantics/ast_symboltable_visitor.cpp [hunk 2]
```diff
@@ -586,19 +587,36 @@ class SymbolTableVisitor : public CommonVisitor<SymbolTableVisitor> {
...
-        process_simd_variables();
         for (size_t i=0; i<x.n_contains; i++) {
             bool current_storage_save = default_storage_save;
             default_storage_save = false;
             visit_program_unit(*x.m_contains[i]);
             default_storage_save = current_storage_save;
         }
+        for (size_t i=0; i<procedure_decl_indices.size(); i++) {
+            try {
+                visit_unit_decl2(*x.m_decl[i]);
```
Comment (kmr-srbh) [inline src/lfortran/semantics/ast_symboltable_visitor.cpp:614]:
```suggestion
        for (size_t i : procedure_decl_indices) {
            try {
                visit_unit_decl2(*x.m_decl[i]);
```

Comment (jinangshah21) [reply src/lfortran/semantics/ast_symboltable_visitor.cpp:614]: Oh yes! Sorry.

**Review by kmr-srbh — APPROVED** [review]
Thanks for this! The changes in the PR look good.

We need to handle this for other language elements that support declarations. Please see `AST.asdl` for nodes with `unit_decl2*`. `AST::Module_t` can be handled in the same way as this PR. 

For others however, like `AST::Subroutine_t`, `AST::Function_t` we probably would need to extend the logic or do something different, as these can have self-referencing procedures. We have an issue open for this at #5735. Another case is `AST::DerivedType_t`, where we declare the procedure in the scope of the derived type, but have the subroutine/function declared outside. I think we can open a separate issue for this.

File: src/lfortran/semantics/ast_symboltable_visitor.cpp [hunk 3]
```diff
@@ -586,19 +587,36 @@ class SymbolTableVisitor : public CommonVisitor<SymbolTableVisitor> {
             }
         }
         for (size_t i=0; i<x.n_decl; i++) {
+            if(x.m_decl[i]->type == AST::unit_decl2Type::Declaration) {
+                AST::Declaration_t decl = (const AST::Declaration_t &)*x.m_decl[i];
```
Comment (certik) [inline src/lfortran/semantics/ast_symboltable_visitor.cpp:591]: Use `AST::down_cast`

File: src/lfortran/semantics/ast_symboltable_visitor.cpp [hunk 4]
```diff
@@ -586,19 +587,36 @@ class SymbolTableVisitor : public CommonVisitor<SymbolTableVisitor> {
...
+                if(decl.m_vartype) {
+                    AST::AttrType_t* type = AST::down_cast<AST::AttrType_t>(decl.m_vartype);
+                    if(type && type->m_type == AST::decl_typeType::TypeProcedure) {
```
Comment (certik) [inline src/lfortran/semantics/ast_symboltable_visitor.cpp:594]: Use `AST::is_a`

**Review by certik — APPROVED** [review]
Otherwise I think it's fine.

File: src/lfortran/semantics/ast_symboltable_visitor.cpp (hunk 4 above)
Comment (jinangshah21) [reply src/lfortran/semantics/ast_symboltable_visitor.cpp:594]: We can't use `is_a` here because we don't have specific node for `TypeProcedure`.

**Review by certik — APPROVED** [review]

**Review by certik — APPROVED** [review]

**Pull request merged**